sql: trim down gossip deprecation calls

tbg opened the pull request:
This series of commits revisits all of the calls to `DeprecatedGossip.Deprecated()` and removes all of them. Instead, a few specialized accessors remain, namely access to the system config and the corresponding trigger channel, as well as resolving between nodeID, storeID, and addresses.

This gives us a precise picture: we need #47150 to avoid the system config gossip, and we need to fix #48432. After that, we *should* be able to start SQL with a nil Gossip (though we would only do so for tenants, due to the many "nonessential" but important things lost when running without Gossip, such as DistSQL).

cc @ajwerner

tbg committed:
changefeedccl: downgrade call to Deprecated

It can be optional since we don't care about CDC working for Phase 2.

Release note: None

tbg committed:
flowinfra: downgrade call to Deprecated()

The use of NodeID is easily replaced with an instance ID.

Release note: None

tbg committed:
sql: downgrade some gossip.Deprecated() calls

Deprecated() calls are blocking phase 2, but the ones
made optional here were not.

Touches #47900.
Touches #47925.

Release note: None

tbg committed:
stmtdiagnostics: avoid calls to gossip.Deprecated

They are now optional.

Release note: None

tbg committed:
stats: downgrade gossip use to optional

Release note: None

tbg committed:
sql: remove gossip.Deprecated() calls from physical planning

None of them were really necessary to enable Phase 2.

Release note: None

tbg committed:
sql: use deprecated getters for sys config

The system config is one main deprecated usage of Gossip. Give it a
special getter to make this easier to see.

Release note: None

tbg committed:
replicaoracle: use narrower gossip interface

This boils down the use of Gossip to the nodeID -> address lookup.
We then remove the `Deprecated()` getter for Gossip.

This means, as of this commit, the only unavoidable use of Gossip in SQL
is for the system config and the replica oracle.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48435)
<!-- Reviewable:end -->

asubiotto reviewed (commented):
> Reviewed 1 of 1 files at r1, 3 of 3 files at r2, 4 of 4 files at r3, 2 of 2 files at r4, 7 of 7 files at r5, 6 of 6 files at r6, 7 of 7 files at r7, 4 of 4 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48435)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @tbg)

___
*[pkg/sql/distsql_physical_planner.go, line 695 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48435#-M6dFqMr-RtF_7Gd-Cyz:-M6dFqMr-RtF_7Gd-Cz-:b8mnwgn) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddff261cc64021203ea14e5688f75a831074e41/pkg/sql/distsql_physical_planner.go#L695)):*
> ```Go
> 
> 	// Check that the node is not draining.
> 	if g, ok := h.gossip.Optional(distsql.MultiTenancyIssueNo); ok {
> ```

I wonder if we should switch these to `Optional` calls to `OptionalErr`. My thinking is that we don't ever want to hit this code in a multitenant configuration. If we're running in a multitenant configuration we should end up with `planCtx.isLocal=true`, which should result in never hitting this code.
___
*[pkg/sql/flowinfra/flow_registry.go, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48435#-M6d9dlyB-LsMAQsjViT:-M6d9dlz8QDhvE9lJCel:b-cbw5lk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c612b28e9ab5820fe29fae846c055a75b377cd3/pkg/sql/flowinfra/flow_registry.go#L108)):*
> ```Go
> 
> 	// instID is the instance ID of the current server. Used for debugging.
> 	instID base.SQLInstanceID
> ```

I'm wondering if we don't just want to remove this altogether. The instance ID is used in error messages that I haven't seen in my lifetime working here.
___
*[pkg/sql/physicalplan/span_resolver_test.go, line 88 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48435#-M6dDPi_1iu-q3IodZBU:-M6dDPi_1iu-q3IodZBV:b-trje39) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddff261cc64021203ea14e5688f75a831074e41/pkg/sql/physicalplan/span_resolver_test.go#L88)):*
> ```Go
> 		s3.Cfg.Settings,
> 		s3.DistSenderI().(*kvcoord.DistSender),
> 		gossip.MakeDeprecatedGossip(s3.Gossip(), true /* exposed */),
> ```

nit: might be nice to add a testing helper that calls`MakeDeprecatedGossip` with `exposed=true`.


<!-- Sent from Reviewable.io -->

tbg committed:
flowinfra: drop instance ID from FlowRegistry

It was only used in an error message that has not shown up in years.

Release note: None

tbg committed:
gossip: improve DeprecatedGossip constructors

Lose the bool and have two constructors instead, one for an exposed
instance and one for the unexposed one.

Release note: None

tbg reviewed (commented):
TFTR, PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48435#-:-M6iOPlNB0HS07gmyttt:b-qs6e22)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @nvanbenschoten)

___
*[pkg/sql/distsql_physical_planner.go, line 695 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48435#-M6dFqMr-RtF_7Gd-Cyz:-M6iHx622eELFmaPDkBF:b-shp4by) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddff261cc64021203ea14e5688f75a831074e41/pkg/sql/distsql_physical_planner.go#L695)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I wonder if we should switch these to `Optional` calls to `OptionalErr`. My thinking is that we don't ever want to hit this code in a multitenant configuration. If we're running in a multitenant configuration we should end up with `planCtx.isLocal=true`, which should result in never hitting this code.
</blockquote></details>

That was my original plan, but then some of them did get hit. I didn't look very carefully, with what you're saying it's likely that they were hit for "silly reasons" that could be avoided. Would you mind taking a look at this when this PR merges? Generally the less branches we have the better.
___
*[pkg/sql/flowinfra/flow_registry.go, line 108 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/48435#-M6d9dlyB-LsMAQsjViT:-M6iHj8e2Tu0Luc-bpRg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c612b28e9ab5820fe29fae846c055a75b377cd3/pkg/sql/flowinfra/flow_registry.go#L108)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I'm wondering if we don't just want to remove this altogether. The instance ID is used in error messages that I haven't seen in my lifetime working here.
</blockquote></details>

Done.
___
*[pkg/sql/physicalplan/span_resolver_test.go, line 88 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48435#-M6dDPi_1iu-q3IodZBU:-M6iOIz3CaRU4IgW7zKJ:b-ic5nba) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddff261cc64021203ea14e5688f75a831074e41/pkg/sql/physicalplan/span_resolver_test.go#L88)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: might be nice to add a testing helper that calls`MakeDeprecatedGossip` with `exposed=true`.
</blockquote></details>

I did something in a similar spirit, see the last commit.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 26 of 26 files at r9, 3 of 3 files at r10, 4 of 4 files at r11, 2 of 2 files at r12, 7 of 7 files at r13, 6 of 6 files at r14, 7 of 7 files at r15, 4 of 4 files at r16, 1 of 1 files at r17, 12 of 12 files at r18.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/48435#-:-M6j-9W65I9wbyUsAAi8:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @nvanbenschoten, and @tbg)

___
*[pkg/sql/distsql_physical_planner.go, line 695 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/48435#-M6dFqMr-RtF_7Gd-Cyz:-M6iybhN-wpm7V-8A0ky:b-80tlqk) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ddff261cc64021203ea14e5688f75a831074e41/pkg/sql/distsql_physical_planner.go#L695)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

That was my original plan, but then some of them did get hit. I didn't look very carefully, with what you're saying it's likely that they were hit for "silly reasons" that could be avoided. Would you mind taking a look at this when this PR merges? Generally the less branches we have the better.
</blockquote></details>

Sure.


<!-- Sent from Reviewable.io -->

tbg commented:
TFTR!

bors r=asubiotto
